ons-spark/spark-concepts/shuffling.md

```diff
@@ -0,0 +1,755 @@
+
+By default, Spark will try and use the broadcast method for joins. We want to ensure that we demonstrate the behaviour of the regular join (sort merge join) and so are deactivating automatic broadcast joins in the configuration by setting `spark.sql.autoBroadcastJoinThreshold` to `-1`.
+````{tabs}
+```{code-tab} py
+spark.stop()
+
+spark = (
+    SparkSession.builder.master("local[2]")
+    .appName("shuffling")
+    # Set number of partitions after a shuffle to 100
+    .config("spark.sql.shuffle.partitions", 100)
+    # Disable broadcast join by default
+    .config("spark.sql.autoBroadcastJoinThreshold", -1)
+    .getOrCreate()
+)
+```
+
+```{code-tab} r R
+
+sparklyr::spark_disconnect(sc)
+
+shuffle_config <- sparklyr::spark_config()
+# Set number of partitions after a shuffle to 100
+shuffle_config$spark.sql.shuffle.partitions <- 100
+# Set number of local partitions to 100
+shuffle_config$spark.sql.local.partitions <- 100
+# Disable broadcast join by default
+shuffle_config$spark.sql.autoBroadcastJoinThreshold <- -1
+
+sc <- sparklyr::spark_connect(
+    master = "local[2]",
+    app_name = "shuffling",
+    config = shuffle_config)
+
+```
+````
+Here we create two DataFrames of $100,000$ rows and add a random number between `1` to `10 `for each called `rand1` and `rand2` respectively, using the same method as previously.
+````{tabs}
+```{code-tab} py
+# Create two DataFrames
+example_2A = spark.range(10**5).withColumn("rand1", F.ceil(F.rand(seed_no) * 10))
+example_2B = spark.range(10**5).withColumn("rand2", F.ceil(F.rand(seed_no + 1) * 10))
+```
+
+```{code-tab} r R
+
+example_2A <- sparklyr::sdf_seq(sc, 0, 10**5 - 1) %>%
+    sparklyr::mutate(rand1 = ceil(rand(seed_no) * 10))
+           
+example_2B <- sparklyr::sdf_seq(sc, 0, 10**5 - 1) %>%
+    sparklyr::mutate(rand2 = ceil(rand(seed_no + 1L) * 10))
+
+```
+````
+We now want to perform some transformations which involve a shuffle. Let's create a new DataFrame, `joined_df`, by joining `example_2B` to `example_2A`. We then group by `rand1` and `rand2`, then sort by `rand1` and `rand2`.
+````{tabs}
+```{code-tab} py
+joined_df = (example_2A.
+             join(example_2B, on="id", how="left")
+             .groupBy("rand1", "rand2")
+             .agg(F.count(F.col("id")).alias("row_count"))
+             .orderBy("rand1", "rand2"))
+```
+
+```{code-tab} r R
+
+joined_df <- example_2A %>%
+    sparklyr::left_join(example_2B, by="id") %>%
+    dplyr::group_by(rand1, rand2) %>%
+    dplyr::summarise(row_count = n()) %>%
+    sparklyr::sdf_sort(c("rand1", "rand2"))
+
+```
+````
+For the action, we are calling a `collect`-style operation (`.toPandas()`/`collect()`) rather than `.show(n)` (PySpark) or `head(n) %>% collect()` (sparklyr). The reason is `.show(n)`/`head(n)` will only do the minimum required to return `n` rows, whereas `.toPandas()`/`collect()` causes all the data to be collected to the driver from the Spark cluster, forcing a full calculation and filling the cache.
```

Comment:
I would avoid using "filling" when saying filling the cache. It will force the full DataFrame to be cached, but the cache capacity of the executor might not be full.

Reply:
I've removed the cache now, so this is no longer relevant.